File: src/main/java/net/volcano/jdacommands/model/command/arguments/implementation/codec/PairCodec.java
package net.volcano.jdacommands.model.command.arguments.implementation.codec;

import kotlin.Pair;
import net.volcano.jdacommands.model.ClassUtil;
import net.volcano.jdacommands.model.ParameterData;
import net.volcano.jdacommands.model.command.arguments.CommandArgument;
import net.volcano.jdacommands.model.command.arguments.implementation.argument.PairArgument;
import net.volcano.jdacommands.model.command.arguments.interfaces.Codec;
import org.springframework.stereotype.Component;

import java.lang.reflect.ParameterizedType;

@Component
public class PairCodec extends Codec<Pair<?, ?>> {
	
	@Override
	protected CommandArgument<Pair<?, ?>> buildArgument(ParameterData data) {
		var builder = PairArgument.builder();
		
		var clazz1 = (Class<?>) ((ParameterizedType) data.parameter.getParameterizedType()).getActualTypeArguments()[0];
		var clazz2 = (Class<?>) ((ParameterizedType) data.parameter.getParameterizedType()).getActualTypeArguments()[1];
		
		if (clazz1.equals(clazz2)) {
			throw new IllegalArgumentException("Pair argument needs to have two different types.");
		}
		
		var codec1 = data.codecRegistry.getCodec((Class<?>) ClassUtil.getCodecType(clazz2));
		var codec2 = data.codecRegistry.getCodec((Class<?>) ClassUtil.getCodecType(clazz2));
		
		if (codec1 == null)
			throw new IllegalArgumentException("Codec for class " + clazz1 + " not found.");
		if (codec2 == null)
			throw new IllegalArgumentException("Codec for class " + clazz2 + " not found.");
		
		var arg1 = codec1.encodeArgument(new ParameterData(
				data.parameter,
				ClassUtil.getActualType(clazz1),
				ClassUtil.getCodecType(clazz1),
				data.codecRegistry
		));
		
		var arg2 = codec2.encodeArgument(new ParameterData(
				data.parameter,
				ClassUtil.getActualType(clazz2),
				ClassUtil.getCodecType(clazz2),
				data.codecRegistry
		));
		
		builder.typeArgument1(arg1);
		builder.typeArgument2(arg2);
		
		return builder.build();
	}
}

File: src/main/java/net/volcano/jdacommands/model/command/arguments/implementation/codec/PairCodec.kt
package net.volcano.jdacommands.model.command.arguments.implementation.codec

import net.volcano.jdacommands.model.ClassUtil.getActualClass
import net.volcano.jdacommands.model.ClassUtil.getCodecClass
import net.volcano.jdacommands.model.ParameterData
import net.volcano.jdacommands.model.command.arguments.CommandArgument
import net.volcano.jdacommands.model.command.arguments.implementation.argument.PairArgument
import net.volcano.jdacommands.model.command.arguments.interfaces.Codec
import org.springframework.stereotype.Component

@Component
class PairCodec : Codec<Pair<*, *>?>() {

	override fun buildArgument(data: ParameterData): CommandArgument<Pair<*, *>?> {

		val clazz1 = data.parameter.type.arguments[0].type!!
		val clazz2 = data.parameter.type.arguments[1].type!!

		require(clazz1 != clazz2) { "Pair argument needs to have two different types." }

		val codec1 = data.codecRegistry.getCodec(getCodecClass(clazz1 as Class<*>))
		val codec2 = data.codecRegistry.getCodec(getCodecClass(clazz2 as Class<*>))

		requireNotNull(codec1) { "Codec for class $clazz1 not found." }
		requireNotNull(codec2) { "Codec for class $clazz2 not found." }

		val arg1 = codec1.encodeArgument(
			ParameterData(
				data.parameter,
				getActualClass(clazz1),
				getCodecClass(clazz1),
				data.codecRegistry
			)
		)

		val arg2 = codec2.encodeArgument(
			ParameterData(
				data.parameter,
				getActualClass(clazz2),
				getCodecClass(clazz2),
				data.codecRegistry
			)
		)

		val pairArg = PairArgument(arg1, arg2)
		pairArg.usage = "<" + arg1.usage + " | " + arg2 + ">"

		return pairArg
	}
}